ehrapy/preprocessing/_bias.py

```diff
@@ -0,0 +1,119 @@
+            of interest. Defaults to 0.1.
+        prediction_confidence_threshold: The threshold for the prediction confidence (R2 or accuracy) of a sensitive feature for predicting another
+            feature to be considered of interest. Defaults to 0.5.
+        corr_method: The correlation method to use. Choose between "pearson" and "spearman". Defaults to "spearman".
+    """
+    from ehrapy.tools import rank_features_supervised
+
+    if sensitive_features == "all":
+        sensitive_features = adata.var_names
+
+    correlations = _feature_correlations(adata, method=corr_method)
+    adata.varp["correlation"] = correlations
+
+    for feature in sensitive_features:
+        for comp_feature in adata.var_names:
+            if correlations.loc[feature, comp_feature] > corr_threshold:
+                logg.warning(
+                    f"Feature {feature} is highly correlated with {comp_feature} (correlation coefficient ≈{correlations.loc[feature, comp_feature]:.3f})."
+                )  # TODO: How do we print results?
+
+    smd_dict = _standardized_mean_differences(adata, sensitive_features)
+    for feature in sensitive_features:
+        abs_smd = smd_dict[feature].abs()
+        for comp_feature in adata.var_names:
+            if abs_smd[comp_feature].max() > smd_threshold:
+                logg.warning(
+                    f"Feature {comp_feature} has a high standardized mean difference with {feature}."
+                )  # TODO: Do we look at / print groups individually?
+
+    for prediction_feature in adata.var_names:
+        prediction_score = rank_features_supervised(
+            adata,
+            prediction_feature,
+            input_features="all",
+            model="rf",
+            key_added=f"{prediction_feature}_feature_importances",
+            percent_output=True,
+            logging=False,
+            return_score=True,
+        )
+        for feature in sensitive_features:
+            feature_importance = adata.var[f"{prediction_feature}_feature_importances"][feature] / 100
+            if feature_importance > feature_importance_threshold and prediction_score > prediction_confidence_threshold:
+                logg.warning(
+                    f"Feature {feature} has a high feature importance for predicting {prediction_feature} (importance in %: {feature_importance:.3f}, prediction score: {prediction_score:.3f})."
+                )
+
+
+def _feature_correlations(adata: AnnData, method: Literal["pearson", "spearman"] = "spearman"):
+    """Computes pairwise correlations between features in the AnnData object.
+
+    Args:
+        adata: An annotated data matrix containing patient data.
+        method: The correlation method to use. Choose between "pearson" and "spearman". Defaults to "spearman".
+
+    Returns:
+        A pandas DataFrame containing the correlation matrix.
+    """
+    corr_matrix = anndata_to_df(adata).corr(method=method)
```

Comment:
When using Spearman's rank correlation coefficient (currently the default), we can calculate it between categorical features as well. While it's probably not be the best statistic for identifying correlations between categorical features, it still gives an overview over potential correlations and biases within categorical and, importantly, also between categorical and continuous features.
Personally, I would keep the categorical features in the correlation matrix. We should still look for alternative statistics better suited for detecting differences in categorical features, though! However, I also think this issue would be at least partially addressed by looking at value counts for categorical features as discussed earlier (e.g., Group a of Feature A is always in Group b of Feature B).

Reply:
Rank correlation only works for ordinal categorical features, and not for nominal categorical variables though..

Cramer's V would likely be what we're looking for here for pairwise categorical-categorical cases of nominal categorical variables: this is the statistic you'd be looking for with your correct intuition for needing value counts by subgroups. A thought here; this does take only values from 0 to 1, and hence might be a bit odd to put into a table with Pearson/spearman Rank Correlation coefficients with values from -1 to 1...

There are multiple cases to take into account though as far as I see, with suggestions in brackets meaning I am not sure and could check out deeper:
- Cont - Cont: Pearson or Spearman
- Categorical Ordinal - Categorical Ordinal: Spearman
- Categorical Nominal - Categorical Nominal: Cramer's V
- Cont - Categorical Nominal: [anything rank_features_groups, ANOVA?]
- Cont - Categorical Ordinal: Spearman
- Categorical Ordinal - categorical Nominal: [anything rank_features_groups, ANOVA?]

Reply:
I've thought about this a bit more, and I think it would be great to use different measurements for the various feature types. However, currently, we don't differentiate between nominal and ordinal categorical features and I'm uncertain if we should incorporate this solely for calculating feature correlations. But maybe there are/will be other use-cases in ehrapy for this? I'm open to adding categorical-ordinal and categorical-nominal feature types as part of addressing issue #701, but I'd like to hear your opinions on whether that makes sense, as it would require the user to carefully review the feature types detected automatically by `ep.ad.infer_feature_types`.

If we choose not to distinguish further between ordinal and nominal categorical features, the best solution I can think of is to use Pearson/Spearman when at least one continuous feature is involved and Cramer's V when we have two categorical features.

Reply:
> However, currently, we don't differentiate between nominal and ordinal categorical features and I'm uncertain if we should incorporate this solely for calculating feature correlations.

I guess because the G-test doesn't care that much whether the categorical data is nominal or ordinal.

> I'm open to adding categorical-ordinal and categorical-nominal feature types as part of addressing issue https://github.com/theislab/ehrapy/issues/701, but I'd like to hear your opinions on whether that makes sense, as it would require the user to carefully review the feature types detected automatically by ep.ad.infer_feature_types.

It should be reasonably easy to detect ordinal categorical data, because that's pretty much what label encoded data looks like anyways. The common case of "0" and "1" would probably be nominal categorical. It's not easy to detect other forms of "nominal categorical" I assume, but that should be fine in most cases.

> if we should incorporate this solely for calculating feature correlations

Since we don't reaaally care about it for G-tests, I'm not sure whether we would make use of it for more things (for now) but I always like being future proof if possible.

TLDR:
1. Can we really detect nominal categorical well? Do we want to put another burden on the user to check this? They'd probably have to label those anyways, because we can't easily detect them (besides 0/1s) and at that point they could label them properly.
2. It doesn't like it's toooo much work to also support storing ordinal vs nominal. I enjoy being future proof.
3. It will increase the code complexity by a non-significant amount because we'd always have to grab both categorical keys/types if we want to apply something to only categorical.

I lean more towards supporting it than not.

Reply:
Just wanted to summarize what I ended up doing to sum up offline and online discussions. As of now, the feature correlations would be the only application in ehrapy where we would need the differentiation between nominal and ordinal features. Because, as soon as data are encoded, detecting this difference automatically is nearly impossible, we would add quite some effort for the user, as they would have to manually declare what features are ordinal and what nominal just to compute the feature correlations using the optimal method (Spearman CC vs. Cramer's V, for instance).

Additionally, just computing Spearman/Pearson CC for all features won't show any correlations that aren't there, but just some correlations between categorical features won't be revealed. However, those should then be detected by the feature importances calculation.

Hence, we decided to stick to Pearson/Spearman CC for all features as of now. If in the future the differentiation between ordinal and nominal categorical features becomes important at other places in ehrapy, it would be easy to adapt the bias detection method accordingly.

Reply:
Thanks for the summary! I would kindly suggest that you create a new issue for the whole topic of "ordinal vs nominal categorical recording" where you paste your well formulated comment. I always search issues for context, but rarely PRs.